include/elements/2d/quadrilateral_bspline_element.tcc

```diff
@@ -96,43 +146,89 @@ inline Eigen::MatrixXd
       pcoord.noalias() +=
           local_shapefn(i) * nodal_coordinates_.row(i).transpose();
 
-    //! Compute the shape function gradient following a multiplicative rule
-    for (unsigned n = 0; n < this->nconnectivity_; ++n)
-      //! Loop over dimension
-      for (unsigned i = 0; i < Tdim; ++i) {
-        double dN_dx = this->gradient(pcoord[i], nodal_coordinates_.row(n)[i],
-                                      (this->node_type_[n])[i], Tpolynomial);
-        switch (this->node_type_[n][i]) {
-          case 1:
-            dN_dx += this->gradient(pcoord[i], nodal_coordinates_.row(n)[i], 5,
+    if (!apply_kernel_correction) {
+      //! Compute the shape function gradient following a multiplicative rule
+      for (unsigned n = 0; n < this->nconnectivity_; ++n)
+        //! Loop over dimension
+        for (unsigned i = 0; i < Tdim; ++i) {
+          double dN_dx = this->gradient(pcoord[i], nodal_coordinates_.row(n)[i],
+                                        (this->node_type_[n])[i], Tpolynomial);
+          switch (this->node_type_[n][i]) {
+            case 1:
+              dN_dx += this->gradient(pcoord[i], nodal_coordinates_.row(n)[i],
```

Comment:
Seems that the kernel correction is just the addition of a new term (for both shape function and gradient), I'm right?

Reply:
Yes for the shape function. The gradient uses a different approach, but similar.